src/util/threadpool.rs

```diff
@@ -0,0 +1,624 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::usize;
+use std::sync::{Arc, Mutex, Condvar};
+use std::thread::{Builder, JoinHandle};
+use std::boxed::FnBox;
+use std::collections::{BinaryHeap, HashMap, VecDeque};
+use std::cmp::Ordering;
+use std::sync::atomic::{AtomicUsize, Ordering as AtomicOrdering};
+use std::hash::Hash;
+use std::marker::PhantomData;
+
+pub struct Task<T> {
+    // The task's number in the pool. Each task has a unique number,
+    // and it's always bigger than preceding ones.
+    id: u64,
+    // which group the task belongs to.
+    group_id: T,
+    task: Box<FnBox() + Send>,
+}
+
+impl<T> Task<T> {
+    fn new<F>(id: u64, group_id: T, job: F) -> Task<T>
+        where F: FnOnce() + Send + 'static
+    {
+        Task {
+            id: id,
+            group_id: group_id,
+            task: Box::new(job),
+        }
+    }
+}
+
+impl<T> Ord for Task<T> {
+    fn cmp(&self, right: &Task<T>) -> Ordering {
+        self.id.cmp(&right.id).reverse()
+    }
+}
+
+impl<T> PartialEq for Task<T> {
+    fn eq(&self, right: &Task<T>) -> bool {
+        self.cmp(right) == Ordering::Equal
+    }
+}
+
+impl<T> Eq for Task<T> {}
+
+impl<T> PartialOrd for Task<T> {
+    fn partial_cmp(&self, rhs: &Task<T>) -> Option<Ordering> {
+        Some(self.cmp(rhs))
+    }
+}
+
+pub trait ScheduleQueue<T> {
+    fn pop(&mut self) -> Option<Task<T>>;
+    fn push(&mut self, task: Task<T>);
+    fn on_task_finished(&mut self, group_id: &T);
+    fn on_task_started(&mut self, group_id: &T);
+}
+
+pub struct PushError<T>(pub T);
+
+// `BigGroupThrottledQueue` tries to throttle group's concurrency to
+//  `group_concurrency_on_busy` when it's busy.
+// When one worker asks a task to run, it schedules in the following way:
+// 1. Find out which group has a running number that is smaller than
+//    that of `group_concurrency_on_busy`.
+// 2. If more than one group meets the first point, run the one who
+//    comes first.
+// If no group meets the first point, choose according to the following rules:
+// 1. Select the groups with least running tasks.
+// 2. If more than one group meets the first point,choose the one
+//     whose task comes first(with the minimum task's ID)
+pub struct BigGroupThrottledQueue<T> {
+    // tasks in pending. tasks in `pending_tasks` have higher priority than
+    // tasks in waiting_queue.
+    pending_tasks: BinaryHeap<Task<T>>,
+    // group_id => tasks array. If `group_concurrency[group_id]` is bigger than
+    // `group_concurrency_on_busy`(which means the number of on-going tasks is
+    // more than `group_concurrency_on_busy`), the rest of the group's tasks
+    // would be pushed into `waiting_queue[group_id]`
+    waiting_queue: HashMap<T, VecDeque<Task<T>>>,
+    // group_id => running_num+pending num. It means there may
```

Comment:
Sorry for the crudeness of previous comment. :)  
It meant to point out that there may be a syntax error for the comment. Try to revise it like  
"It means at most `group_concurrency[group_id]` tasks of the group may be running."